internal/status/advl4_status.go

```diff
@@ -221,38 +248,104 @@ func UpdateGatewayStatusObject(gw *advl4v1alpha1pre1.Gateway, updateStatus *advl
-	for _, listener := range gw.Spec.Listeners {
-		listenerStatuses = append(listenerStatuses, advl4v1alpha1pre1.ListenerStatus{
-			Port: strconv.Itoa(int(listener.Port)),
-			Conditions: []advl4v1alpha1pre1.ListenerCondition{{
-				Type:               "Ready",
-				Status:             corev1.ConditionFalse,
-				LastTransitionTime: metav1.Now(),
-			}},
-		})
+	for _, listener := range gwSpec.Listeners {
+		if val, ok := thisMap[strconv.Itoa(int(listener.Port))]; ok {
+			listenerStatuses = append(listenerStatuses, advl4v1alpha1pre1.ListenerStatus{
+				Port:       strconv.Itoa(int(listener.Port)),
+				Conditions: val,
+			})
+		} else {
+			var portCondition corev1.ConditionStatus
+			if gwReady {
+				portCondition = corev1.ConditionTrue
+			} else {
+				portCondition = corev1.ConditionFalse
+			}
+			listenerStatuses = append(listenerStatuses, advl4v1alpha1pre1.ListenerStatus{
+				Port: strconv.Itoa(int(listener.Port)),
+				Conditions: []advl4v1alpha1pre1.ListenerCondition{{
+					Type:               "Ready",
+					Status:             portCondition,
+					LastTransitionTime: metav1.Now(),
+				}},
+			})
+		}
 	}
+
 	gwStatus.Listeners = listenerStatuses
 	if len(gwStatus.Addresses) == 0 {
 		gwStatus.Addresses = []advl4v1alpha1pre1.GatewayAddress{}
 	}
+	return
+}
+
+// getGateways fetches all ingresses and returns a map: {"namespace/name": ingressObj...}
+// if bulk is set to true, this fetches all ingresses in a single k8s api-server call
+func getGateways(gwNSNames []string, bulk bool, retryNum ...int) map[string]*advl4v1alpha1pre1.Gateway {
+	retry := 0
+	gwMap := make(map[string]*advl4v1alpha1pre1.Gateway)
+	if len(retryNum) > 0 {
+		utils.AviLog.Infof("msg: Retrying to get the gateway for status update")
+		retry = retryNum[0]
+		if retry >= 2 {
+			utils.AviLog.Errorf("msg: getGateways for status update retried 3 times, aborting")
+			return gwMap
+		}
+	}
+
+	if bulk {
+		gwList, err := lib.GetAdvL4Clientset().NetworkingV1alpha1pre1().Gateways("").List(metav1.ListOptions{})
+		if err != nil {
+			utils.AviLog.Warnf("Could not get the gateway object for UpdateStatus: %s", err)
+			// retry get if request timeout
+			if strings.Contains(err.Error(), utils.K8S_ETIMEDOUT) {
+				return getGateways(gwNSNames, bulk, retry+1)
+			}
+		}
+		for i := range gwList.Items {
+			ing := gwList.Items[i]
+			gwMap[ing.Namespace+"/"+ing.Name] = &ing
+		}
+		return gwMap
+	}
+
+	for _, namespaceName := range gwNSNames {
+		nsNameSplit := strings.Split(namespaceName, "/")
+		gw, err := lib.GetAdvL4Clientset().NetworkingV1alpha1pre1().Gateways(nsNameSplit[0]).Get(nsNameSplit[1], metav1.GetOptions{})
+		if err != nil {
+			utils.AviLog.Warnf("msg: Could not get the gateway object for UpdateStatus: %s", err)
+			// retry get if request timeout
+			if strings.Contains(err.Error(), utils.K8S_ETIMEDOUT) {
+				return getGateways(gwNSNames, bulk, retry+1)
+			}
+		}
+
+		gwMap[gw.Namespace+"/"+gw.Name] = gw
```

Comment:
seems like if we get an error, then also we will hit this code block and nil pointer would be seen.